Add Attribute::getOptional<T>() and use to add some more dynamic datatype conversions at read time

franzpoeschel opened the pull request:
Currently, we have `Attribute::get<T>()` that throws a `std::runtime_error()` if no conversion is possible.
Adding `Attribute::getOptional<T>()` is helpful to check if a conversion is possible without having to resort to catching exceptions.

This PR also uses that new API method while reading a Series in. This became necessary in #1277 from where I have extracted this PR (that PR adds a less verbose mode to the JSON backend that does not explicitly annotate datatypes and might hence report different numerical types at read times).

franzpoeschel committed:
Add and use Attribute::getOptional<T>()

Useful for backends that might not report back the right numerical type

franzpoeschel commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -251,4 +370,16 @@ U Attribute::get() const
     return getCast<U>(Variant::getResource());
```
We could remove `getCast` here in a similar way. Slightly API-breaking though since this is in a public header.

franzpoeschel replied:
I did this API break now. All implementation details are now moved to the `detail` namespace.

franzpoeschel committed:
CI fixes, clean up append_test

franzpoeschel commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -226,6 +241,110 @@ auto doConvert(T *pv) -> U
 }
 #endif
 
+template <typename T, typename U>
```
Use `std::variant<T, std::string>` to combine both functions into one and still have error messages

franzpoeschel replied:
done

ax3l commented on include/openPMD/backend/Attribute.hpp:
```diff
@@ -117,6 +118,20 @@ class Attribute
+     *          An empty std::optional if no conversion is possible.
+     */
+    template <typename U>
+    std::optional<U> getOptional() const;
```
Please add in Python, too (pybind11 returns `py:None` for `std::nullopt`)

franzpoeschel replied:
Python does not use this call at all:
```c++
        // C++ pass-through API: Getter
        .def(
            "get_attribute",
            [](Attributable &attr, std::string const &key) {
                auto v = attr.getAttribute(key);
                return v.getResource();
                // TODO instead of returning lists, return all arrays (ndim > 0)
                // as numpy arrays?
            })
```

There is no explicit type requesting as `get<T>()` does, in Python you just get the type that there is

ax3l replied:
Yes, I was wondering if we want to add such an option in Python to cast?

franzpoeschel committed:
Unify code paths for both get() and getOptional()

Note that this is slightly API breaking since the shared code was in a
public header.
All implementation details are now in the detail namespace.
